subsys/mgmt/mcumgr/smp/src/mcumgr_client.c

```diff
@@ -0,0 +1,713 @@
+
+const char smp_images_str[] = "images";
+
+static int echo_res_fn(struct net_buf *nb, void *user_data)
+{
+	struct mcumgr_client_os_gr_rsp response;
+	zcbor_state_t zsd[4];
+	struct zcbor_string value = {0};
+	size_t decoded;
+	bool ok;
+
+	response.cmd = MGMT_OS_ECHO_RSP;
+	response.echo.echo_res = NULL;
+	response.echo.echo_rsp_len = 0;
+
+	if (!nb) {
+		LOG_ERR("Echo command timeout");
+		response.status = MGMT_ERR_ETIMEOUT;
+		client_gr_os_fn(&response);
+		k_sem_give(user_data);
+		return 0;
+	}
+
+	struct zcbor_map_decode_key_val echo_res_decode[] = {
+		ZCBOR_MAP_DECODE_KEY_VAL(r, zcbor_tstr_decode, &value)};
+
+	/* Init ZCOR decoder state */
+	zcbor_new_decode_state(zsd, ARRAY_SIZE(zsd), nb->data, nb->len, 1);
+
+	ok = zcbor_map_decode_bulk(zsd, echo_res_decode, ARRAY_SIZE(echo_res_decode), &decoded) ==
+	     0;
+
+	if (!ok || value.len == 0) {
+		response.status = MGMT_ERR_ECORRUPT;
+		client_gr_os_fn(&response);
+		k_sem_give(user_data);
+		return MGMT_ERR_ECORRUPT;
+	}
+
+	response.echo.echo_res = value.value;
+	response.echo.echo_rsp_len = value.len;
+	response.status = MGMT_ERR_EOK;
+	client_gr_os_fn(&response);
+	k_sem_give(user_data);
+	return 0;
+}
+
+static int reset_res_fn(struct net_buf *nb, void *user_data)
+{
+	struct mcumgr_client_os_gr_rsp response;
+
+	if (!nb) {
+		response.status = MGMT_ERR_ETIMEOUT;
+	} else {
+		response.status = MGMT_ERR_EOK;
+	}
+	client_gr_os_fn(&response);
+	k_sem_give(user_data);
+	return 0;
+}
+
+static bool validate_parsed_list_data(struct zcbor_map_decode_key_val *map, size_t map_size,
+				      struct mcumgr_image_data *src_image,
+				      struct img_mgmt_list_res *image_data)
+{
+	if (map->found) {
+		src_image->img_num = image_data->image;
+	} else {
+		src_image->img_num = 0;
+	}
+
+	map++;
+	for (int i = 1; i < map_size; i++) {
+		if (!map->found && i < 4) {
+			return false;
+		}
+		switch (i) {
+		case 1:
+			src_image->slot_num = image_data->slot;
+			break;
+		case 2:
+			if (image_data->version.len == 0 ||
+			    image_data->version.len >= sizeof(src_image->version)) {
+				return false;
+			}
+			memcpy(src_image->version, image_data->version.value,
+			       image_data->version.len);
+			src_image->version[image_data->version.len] = '\0';
+			break;
+		case 3:
+			if (image_data->hash.len == 0 ||
+			    image_data->hash.len >= sizeof(src_image->hash)) {
+				return false;
+			}
+			memcpy(src_image->hash, image_data->hash.value, image_data->hash.len);
+			src_image->hash[image_data->hash.len] = '\0';
+			break;
+		case 4:
+			src_image->bootable = map->found ? image_data->bootable : false;
+			break;
+		case 5:
+			src_image->pending = map->found ? image_data->pending : false;
+			break;
+		case 6:
+			src_image->confirmed = map->found ? image_data->confirmed : false;
+			break;
+		case 7:
+			src_image->active = map->found ? image_data->active : false;
+			break;
+		case 8:
+			src_image->permanent = map->found ? image_data->permanent : false;
+			break;
+		}
+		map++;
```

Comment:
I think this is al little bit too complicated. Take look at `ZCBOR_MAP_DECODE_KEY_VAL`: it takes `zcbor_decoder_t ` signature type function and last parameter is destination for decoded value.
If you define function like
```
bool get_confirmed(zcbor_state_t *state, struct image_data *out)
{
    bool val;
    bool fin = zcbor_bool_decode(state, &val);
    if (!fin) {
       return false;
    }
    // here goes val validation and assignment to image_data 
   ...
    return true;
}
```

you can avoid  looping over the map to validate data, you can validate it as it comes. You can even break out from parsing loop if you return false here. Unfortunately you will have to store some additional failure reason somewhere within your `*out` pointed data structure, but that makes sense because you do not want to mix bulk decoder and validation errors.

Reply:
Good point. I like this idea.